src/analyzer.ts

```diff
@@ -8,762 +8,153 @@
-      var uri = url.resolve(href, scriptHref);
-      var content = this._content[uri];
-      var inlined = dom5.constructors.element('script');
-      dom5.setTextContent(inlined, '\n' + content + '\n');
-      dom5.replace(script, inlined);
-    });
-    return scripts.length > 0;
-  };
+    return new DocumentDescriptor(document, entities);
+  }
 
-  _inlineImports(ast:dom5.Node, href:string, loaded:{[href:string]:boolean}) {
-    var imports = dom5.queryAll(ast, isHtmlImportNode);
-    imports.forEach((htmlImport) => {
-      var importHref = dom5.getAttribute(htmlImport, 'href');
-      var uri = url.resolve(href, importHref);
-      if (loaded[uri]) {
-        dom5.remove(htmlImport);
-        return;
-      }
-      var content = this.getLoadedAst(uri, loaded);
-      dom5.replace(htmlImport, content);
-    });
-    return imports.length > 0;
-  };
 
   /**
-   * Returns a promise resolving to a form of the AST with all links replaced
-   * with the document they link to. .css and .script files become &lt;style&gt; and
-   * &lt;script&gt;, respectively.
-   *
-   * The elements in the loaded document are unmodified from their original
-   * documents.
-   *
-   * @param {string} href The document to load.
-   * @param {Object.<string,boolean>=} loaded An object keyed by already loaded documents.
-   * @return {Promise.<DocumentAST>}
+   * Loads and parses a single file, deduplicating any requrests for the same
+   * URL.
    */
-  getLoadedAst(href:string, loaded:{[href:string]:boolean}) {
-    if (!loaded) {
-      loaded = {};
+  async load(url: string): Promise<Document<any, any>> {
+    // TODO(justinfagnani): normalize url
+    if (this._documents.has(url)) {
+      return this._documents.get(url);
     }
-    loaded[href] = true;
-    var parsedDocument = this.parsedDocuments[href];
-    var analyzedDocument = this.html[href];
-    var astCopy = dom5.parse(dom5.serialize(parsedDocument));
-    // Whenever we inline something, reset inlined to true to know that anoather
-    // inlining pass is needed;
-    this._inlineStyles(astCopy, href);
-    this._inlineScripts(astCopy, href);
-    this._inlineImports(astCopy, href, loaded);
-    return astCopy;
-  };
-
-  /**
-   * Calls `dom5.nodeWalkAll` on each document that `Anayzler` has laoded.
-   */
-  nodeWalkDocuments(predicate:dom5.Predicate) {
-    var results: dom5.Node[] = [];
-    for (var href in this.parsedDocuments) {
-      var newNodes = dom5.nodeWalkAll(this.parsedDocuments[href], predicate);
-      results = results.concat(newNodes);
+    if (!this._loader.canLoad(url)) {
+      throw new Error(`Can't load URL: ${url}`);
     }
-    return results;
-  };
+    // Use an immediately executed async function to create the final Promise
+    // synchronously so we can store it in this._documents before any other
+    // async operations to avoid any race conditions.
+    let promise = (async () => {
+      let content = await this._loader.load(url);
```

Comment:
`this._loader.load(url)` is called before `this._documents.set(url, promise)` is.

I confirmed this with the following script:

```javascript
console.log('first');
const p = (async () => {
  console.log('second');
  await Promise.resolve();
  console.log('fourth');
})();
console.log('third');
```

I think to be confident that this is correct we should start the function with `await Promise.resolve()`